RDI-2895 RDI-2873 Save recon in bsinfo

marecki315 opened the pull request:
Access to the reconstructed frame from the H.264 encoder is useful in scenarios where the application layer needs to perform further processing or analysis on the encoded video. For example, in an LCEVC (Low Complexity Enhancement Video Coding) encoder, the reconstructed frame is required to generate enhancement layers that improve visual quality. By exposing the reconstructed picture, the encoder facilitates integration with advanced video processing workflows and enables more flexible application-level control.

msabiniok commented on codec/api/wels/codec_app_def.h:
```diff
@@ -651,6 +651,9 @@ typedef struct {
   int iFrameSizeInBytes;
   long long uiTimeStamp;
+  unsigned char*  pReconData[3];   ///< plane data of reconstructed frame
+  bool      bSaveRecon;            ///< save reconstructed frame to pReconData
```
I like this solution better then copyLastReconFrame. Nevertheless i would avoid putting bSaveRecon here as SFrameBSInfo is an output structure and this would need to be set before passing it to encode function as input parameter. I would see it actuallt with ENCODER_OPTION_ENABLE_RECON_FRAME  option

msabiniok replied:
It also affects the least the code of openh.

ylatuya replied:
I believe the correct place to configure this is in the input structure `SSourcePicture`, similar to how it's done to enable PSNR computation:

```
  bool      bPsnrY;                ///< get Y PSNR for this frame
  bool      bPsnrU;                ///< get U PSNR for this frame
  bool      bPsnrV;                ///< get V PSNR for this frame
 ```
 
We should check if PSNR calculations are done in-place or if they are also generating a recon frame to avoid creating 2.

ylatuya replied:
@marecki315 I think this comment was not process. In the current PR I don't understand how the user can control the creation of the recon frame

ylatuya commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3907,44 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
                  "WelsEncoderEncodeExt()MinCr Checking,codec bitstream size is larger than Level limitation");
     }
+
+    if (iSpatialNum == 1 && pFbi->bSaveRecon && NULL != pFbi->pReconData[0]) {
```
We must check that pReconData has been correctly initialized or allocate memory for the recon frame otherwise.

ylatuya commented on codec/api/wels/codec_app_def.h:
```diff
@@ -651,6 +651,9 @@ typedef struct {
   EVideoFrameType eFrameType;
   int iFrameSizeInBytes;
   long long uiTimeStamp;
+  unsigned char*  pReconData[3];   ///< plane data of reconstructed frame
```
I think we should either reuse `SSourcePicture` or define a new structure that can provide the correct information about colorspace, size, planes and strides for the recon frame.

marecki315 committed:
Enable access to reconstructed picture output in the encoder

Access to the reconstructed frame from the H.264 encoder is useful in
scenarios where the application layer needs to perform further
processing or analysis on the encoded video. By exposing the
reconstructed picture, the encoder facilitates integration with
advanced video processing workflows and enables more flexible
application-level control.

msabiniok commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3908,57 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
+      }
+
+      // Copy U and V planes
+      for (int i = 1; i < 3; ++i) {
```
I do not know if it may happen but this code takes into account that someone may want only luma, if someone would provide NULL for luma but a valid pointer for chroma, it will fall into HaveRecon=false.
On the other hand, there is Andoni comment on allocation in case of not allocated memory by user.

msabiniok replied:
In case of VNOVA SDK, the specific plane is requested by the SDK as it provides ReconFrame pointers i.ex [Luma, NULL, NULL]. Then if we provide a recon array from VNOVA and internally allocate for U/V planes. It would be at least confusing, and possibly not properly cleaned up.

msabiniok commented on codec/encoder/plus/src/welsEncoderExt.cpp:
```diff
@@ -1184,6 +1184,13 @@ int CWelsH264SVCEncoder::SetOption (ENCODER_OPTION eOptionId, void* pOption) {
              "CWelsH264SVCEncoder::SetOption():ENCODER_OPTION_BITS_VARY_PERCENTAGE,iBitsVaryPercentage = %d", iValue);
   }
+  break;
+    case ENCODER_OPTION_ENABLE_RECON_PIC: {
```
I think this could be supplemental with having bReconPic in Srource picture as Andoni mentioned. So it could either always on with this as in OPTION_DUMP_RECON, or for specific frame when bRecoPic is set?

ylatuya commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3907,61 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
+            || NULL != pFbi->pReconPic->pData[2])) {
+      SReconPicture* pReconPic = pFbi->pReconPic;
+      unsigned char* pReconData[4] = {pReconPic->pData[0], pReconPic->pData[1],
+                                      pReconPic->pData[2], pReconPic->pData[3]};
```
What is this 4th pointer in the pData array?

ylatuya commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3907,61 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
                  "WelsEncoderEncodeExt()MinCr Checking,codec bitstream size is larger than Level limitation");
     }
+
+    if (iSpatialNum == 1 && NULL != pFbi->pReconPic
```
I would move all this code block into a new function.

ylatuya commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3907,61 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
     }
+
+    if (iSpatialNum == 1 && NULL != pFbi->pReconPic
+        && (NULL != pFbi->pReconPic->pData[0]
```
I would also invert these checks to return early end avoid an extra indentation:

```
if (iSpatialNum != 1 || NULL == pFbi->pReconPic || 
       (NULL == pFbi->pReconPic->pData[0] &&
        NULL == pFbi->pReconPic->pData[1] &&
        NULL == pFbi->pReconPic->pData[2]) {
     return;
 }
```

ylatuya reviewed (approved):
To upstream this, it will also need tests and probably update documentation somewhere or update the samples

ylatuya commented on codec/encoder/core/src/encoder.cpp:
```diff
@@ -544,7 +544,61 @@ void DumpRecFrame (SPicture* pCurPicture, const char* kpFileName, const int8_t k
+ * \brief   Copy the reconstructed picture to SFrameBSInfo structure
+ */
+void CopyReconPicToBSInfo (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, SPicture* pSrcRecPic, const SSourcePicture* pSrcPic) {
+  if (NULL != pFbi->pReconPic->pData[0] || NULL != pFbi->pReconPic->pData[1] || NULL != pFbi->pReconPic->pData[2]) {
```
This is more a personal liking by I prefer early returns since they improve readabilidy of code:

```
if (NULL = pFbi->pReconPic->pData[0] && NULL == pFbi->pReconPic->pData[1] || NULL == pFbi->pReconPic->pData[2]) {
  pFbi->bHaveRecon = false;
  return;
}

/* Do the actual work */

 ```

ylatuya commented on codec/encoder/core/src/encoder_ext.cpp:
```diff
@@ -3907,6 +3907,10 @@ int32_t WelsEncoderEncodeExt (sWelsEncCtx* pCtx, SFrameBSInfo* pFbi, const SSour
                  "WelsEncoderEncodeExt()MinCr Checking,codec bitstream size is larger than Level limitation");
     }
+
+    if (iSpatialNum == 1 && NULL != pFbi->pReconPic)
```
Can we keep these checks inside the `CopyReconPicToBSInfo` functions? We need to ensure that `bHaveRecon` is to `false` since we don't initialize this structure.
